libraries/ESP8266WebServer/src/ESP8266WebServer.cpp

```diff
@@ -121,15 +127,97 @@ bool ESP8266WebServer::authenticate(const char * username, const char * password
+        authReq = String();
+        return false;
+      }
+      // extracting required parameters for RFC 2069 simpler Digest
+      String _realm    = _exractparam(authReq,"realm=\"");
+      String _nonce    = _exractparam(authReq,"nonce=\"");
+      String _uri      = _exractparam(authReq,"uri=\"");
+      String _response = _exractparam(authReq,"response=\"");
+      String _opaque   = _exractparam(authReq,"opaque=\"");
+
+      if((!_realm.length())||(!_nonce.length())||(!_uri.length())||(!_response.length())||(!_opaque.length())){
+        authReq = String();
+        return false;
+      }
+      if((_opaque!=_sopaque)||(_nonce!=_snonce)||(_realm!=_srealm)){
+        authReq = String();
+        return false;
+      }
+      // parameters for the RFC 2617 newer Digest
+      String _nc,_cnonce;
+      if(authReq.indexOf("qop=auth") != -1){
+        _nc = _exractparam(authReq,"nc=",',');
+        _cnonce = _exractparam(authReq,"cnonce=\"");
+      }
+      MD5Builder md5;
+      md5.begin();
+      md5.add(String(username)+":"+_realm+":"+String(password));  // md5 of the user:realm:user
+      md5.calculate();
+      String _H1 = md5.toString();
+      #ifdef DEBUG_ESP_HTTP_SERVER
+      DEBUG_OUTPUT.println("Hash of user:realm:pass=" + _H1);
+      #endif
+      md5.begin();
+      if(_currentMethod == HTTP_GET){
+        md5.add("GET:"+_uri);
+      }else if(_currentMethod == HTTP_POST){
+        md5.add("POST:"+_uri);
+      }else if(_currentMethod == HTTP_PUT){
+        md5.add("PUT:"+_uri);
+      }else if(_currentMethod == HTTP_DELETE){
+        md5.add("DELETE:"+_uri);
+      }else{
+        md5.add("GET:"+_uri);
+      }
+      md5.calculate();
+      String _H2 = md5.toString();
+      #ifdef DEBUG_ESP_HTTP_SERVER
+      DEBUG_OUTPUT.println("Hash of GET:uri=" + _H2);
+      #endif
+      md5.begin();
+      if(authReq.indexOf("qop=auth") != -1){
+        md5.add(_H1+":"+_nonce+":"+_nc+":"+_cnonce+":auth:"+_H2);
+      }else{
+        md5.add(_H1+":"+_nonce+":"+_H2);
+      }
+      md5.calculate();
+      String _responsecheck = md5.toString();
+      #ifdef DEBUG_ESP_HTTP_SERVER
+      DEBUG_OUTPUT.println("The Proper response=" +_responsecheck);
+      #endif
+      if(_response==_responsecheck){
+        authReq = String();
+        return true;
+      }
     }
     authReq = String();
   }
   return false;
 }
 
-void ESP8266WebServer::requestAuthentication(){
-  sendHeader("WWW-Authenticate", "Basic realm=\"Login Required\"");
-  send(401);
+void ESP8266WebServer::requestAuthentication(HTTPAuthMethod mode, const char* realm, const String& AuthFailMsg){
+  if(realm==NULL){
+    _srealm = "Login Required";
+  }else{
+    _srealm = String(realm);
+  }
+  if(mode==BASIC_AUTH){
+    sendHeader("WWW-Authenticate", "Basic realm=\"" + _srealm + "\"");
+  }else{
+    _snonce=String(RANDOM_REG32,HEX)+String(RANDOM_REG32,HEX)+String(RANDOM_REG32,HEX)+String(RANDOM_REG32,HEX);
```

Comment:
Does String(val, HEX) automatically pad the string to 8 hexadecimal places, or the string can end up being shorter?

Reply:
Also, it might be a bit nicer to put this into a separate function, a la `String getRandomHexString(size_t length)`.

Reply:
String(val, HEX) in the original Arduino depends on itoa which does not support padding
But it's not a problem really
the nonce and opaque values are arbitrary in length anyway

but the idea of a function is much better as I need to call it twice ..
I will modify it right away ..